base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/generation/GeneratorContext.java

```diff
@@ -22,165 +32,154 @@
  *
  */
 public class GeneratorContext extends ModelObject {
-	
-	/**
-	 * The property to listen for changes in a Generator containing the mapping Map<Integer, String> of
-	 *  validity checks for each row of the script generator.
-	 */
-	private static final String VALIDITY_ERROR_MESSAGE_PROPERTY = "validity error messages";
-	
-	/**
-	 * The property to listen for changes in a Generator containing whether or not all
-	 *  script generator contents are valid or not (bool).
-	 */
-	private static final String PARAM_VALIDITY_PROPERTY = "parameter validity";
-	
-	/**
-	 * The property to listen for changes in a Generator containing the generated script (String).
-	 */
-	private static final String GENERATED_SCRIPT_PROPERTY = "generated script";
+	private static final Logger LOG = IsisLog.getLogger(ScriptGeneratorSingleton.class);
 	
 	/**
 	 * A mapping from the type of generated language to it's generator.
 	 */
-	private Map<GeneratedLanguage, AbstractGenerator> generatorStrategies = new HashMap<>();
+	private Map<GenerationType, Strategy> generatorStrategies = new HashMap<>();
 	
 	
 	/**
-	 * Add a generator to use (will replace the languages generator with a new one if it matches).
-	 * Listen to changes in the generators validity checking and script generation.
-	 * 
-	 * @param generatedLanguage The language the new generator uses.
-	 * @param generator The generator to use.
+	 * Add listener to the generators present in this context
+	 * @param propertyName property to listen to
+	 * @param type is it a programming language generator or data exchange format file generator?
 	 */
-	@SuppressWarnings("unchecked")
-	public void putGenerator(GeneratedLanguage generatedLanguage, AbstractGenerator generator) {
-		generator.addPropertyChangeListener(VALIDITY_ERROR_MESSAGE_PROPERTY, evt -> {
-			firePropertyChange(VALIDITY_ERROR_MESSAGE_PROPERTY, evt.getOldValue(), evt.getNewValue());
-		});
-		generator.addPropertyChangeListener(PARAM_VALIDITY_PROPERTY, evt -> {
-			firePropertyChange(PARAM_VALIDITY_PROPERTY, evt.getOldValue(), evt.getNewValue());
-		});
-		generator.addPropertyChangeListener(GENERATED_SCRIPT_PROPERTY, evt -> {
-			firePropertyChange(GENERATED_SCRIPT_PROPERTY, null, evt.getNewValue());
-		});
-		generatorStrategies.put(generatedLanguage, generator);
+	public void AddListenerToGenerator(String propertyName, GenerationType type) {
+		try {
+			if (type.getType() == GenerationType.PROGRAMMING_LANGUAGE) {
+				AbstractProgrammingLanguageGenerator generator = (AbstractProgrammingLanguageGenerator)getGenerator(type);
```

Comment:
As I am suggesting to split Generator and Parameter loading and saving this can be switched back to putGenerator

Reply:
removed.